File: apkEditor/src/main/java/com/mcal/apkeditor/activities/OdexPatchActivity.java
package com.mcal.apkeditor.activities;

import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatButton;
import androidx.appcompat.widget.AppCompatEditText;
import androidx.appcompat.widget.Toolbar;

import com.mcal.apkeditor.R;
import com.mcal.apkeditor.dialogs.FileSelectDialog;
import com.mcal.apkeditor.dialogs.ProcessingDialog;
import com.mcal.apkeditor.utils.OdexPatcher;
import com.mcal.common.utils.ApkInfoParser;

import org.jetbrains.annotations.NotNull;

/**
 * Created by phe3 on 1/30/2018.
 */

public class OdexPatchActivity extends AppCompatActivity implements View.OnClickListener, FileSelectDialog.IFileSelection {
    private AppCompatEditText apkPathEt;
    private String apkPath;

    @Override
    public void onCreate(Bundle bundle) {
        super.onCreate(bundle);

        setContentView(R.layout.activity_odex_patch);
        setupToolbar("Odex Patcher");

        initView();
    }

    private void setupToolbar(String title) {
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle(title);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
    }

    private void initView() {
        this.apkPathEt = (AppCompatEditText) findViewById(R.id.et_apkpath);

        AppCompatButton selectBtn = (AppCompatButton) findViewById(R.id.btn_select_apkpath);
        selectBtn.setOnClickListener(this);
        AppCompatButton applyBtn = (AppCompatButton) findViewById(R.id.btn_apply_patch);
        applyBtn.setOnClickListener(this);
    }

    @Override
    public void onClick(@NonNull View v) {
        int id = v.getId();
        if (id == R.id.btn_select_apkpath) {
            new FileSelectDialog(this, this, ".apk", "", null);
        } else if (id == R.id.btn_apply_patch) {
            this.apkPath = apkPathEt.getText().toString();
            ProcessingDialog dlg = new ProcessingDialog(this, new PatchProcessor(), -1);
            dlg.show();
        }
    }

    ////////////////////////////////////////////////////////////////////////////////
    // Processing Dialog

    @Override
    public void fileSelectedInDialog(String filePath, String extraStr, boolean openFile) {
        apkPathEt.setText(filePath);
    }

    ////////////////////////////////////////////////////////////////////////////////
    // File selection

    @Override
    public boolean isInterestedFile(@NonNull String filename, String extraStr) {
        return filename.endsWith(".apk");
    }

    @Override
    public String getConfirmMessage(String filePath, String extraStr) {
        return null;
    }

    ////////////////////////////////////////////////////////////////////////////////
    @Override
    public boolean onOptionsItemSelected(@NotNull MenuItem item) {
        // Respond to the action bar's Up/Home button
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    class PatchProcessor implements ProcessingDialog.ProcessingInterface {
        private String errMessage;
        private String odexPath;

        @Override
        public void process() throws Exception {
            ApkInfoParser parser = new ApkInfoParser();
            ApkInfoParser.AppInfo info = parser.parse(OdexPatchActivity.this, apkPath);
            if (info == null) {
                return;
            }

            String packageName = info.packageName;
            OdexPatcher patcher = new OdexPatcher(packageName);
            patcher.applyPatch(OdexPatchActivity.this, apkPath);

            odexPath = patcher.targetOdex;
            if (patcher.errMessage != null) {
                this.errMessage = patcher.errMessage;
                throw new Exception(errMessage);
            }
        }

        @Override
        public void afterProcess() {
            if (errMessage == null) {
                Toast.makeText(OdexPatchActivity.this, "Patched to " + odexPath, Toast.LENGTH_LONG).show();
            }
        }
    }
}

File: apkEditor/src/main/java/com/mcal/apkeditor/activities/OdexPatchActivity.kt
package com.mcal.apkeditor.activities

import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import com.mcal.apkeditor.R
import com.mcal.apkeditor.databinding.ActivityOdexPatchBinding
import com.mcal.apkeditor.utils.OdexPatcher
import com.mcal.common.activities.CustomizedLangActivity
import com.mcal.common.filesystem.FilePickHelper
import com.mcal.common.utils.ApkInfoParser
import com.mcal.common.utils.ScopedStorage
import com.mcal.common.utils.copyFile
import com.mcal.common.view.ProgressDialog
import com.mcal.common.view.ProgressDialog.ProcessingInterface
import java.io.File

/**
 * Created by phe3 on 1/30/2018.
 */
class OdexPatchActivity : CustomizedLangActivity() {
    private lateinit var binding: ActivityOdexPatchBinding
    private lateinit var pickLauncher: ActivityResultLauncher<Intent>
    private var apkPath: String? = null

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityOdexPatchBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setupToolbar(R.id.toolbar, "Odex Patcher", back = true)
        initView()
    }

    private fun initView() {
        binding.btnSelectApkpath.setOnClickListener {
            pickApk()
        }
        binding.btnApplyPatch.setOnClickListener {
            apkPath?.takeIf { it.isNotEmpty() && File(it).exists() }?.let {
                ProgressDialog(
                    this@OdexPatchActivity, "", "Working…", false,
                    PatchProcessor(), -1
                ).show()
            } ?: run {
                Toast.makeText(this, R.string.msg_unsupported_file, Toast.LENGTH_SHORT).show()
            }
        }
        pickLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == RESULT_OK) {
                result.data?.data?.let { uri ->
                    val apk = File(ScopedStorage.getTmpDir().path, FilePickHelper.getFileName(this, uri))
                    contentResolver.openInputStream(uri)?.let { bytes -> copyFile(bytes, apk) }
                    if (apk.exists() && apk.name.endsWith(".apk")) {
                        val path = apk.path.also { apkPath = it }
                        binding.etApkpath.setText(path)
                    } else {
                        Toast.makeText(this, R.string.msg_unsupported_file, Toast.LENGTH_SHORT).show()
                    }
                }
            }
        }
    }

    private fun pickApk() {
        pickLauncher.launch(FilePickHelper.pickFile(true))
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    internal inner class PatchProcessor : ProcessingInterface {
        private var errMessage: String? = null
        private var odexPath: String? = null

        @Throws(Exception::class)
        override fun process() {
            val parser = ApkInfoParser()
            val info = parser.parse(this@OdexPatchActivity, apkPath) ?: return
            val packageName = info.pkgName
            val patcher = OdexPatcher(packageName)
            patcher.applyPatch(this@OdexPatchActivity, apkPath)
            odexPath = patcher.targetOdex
            patcher.errMessage?.let { msg ->
                errMessage = msg
                throw Exception(errMessage)
            }
        }

        override fun afterProcess() {
            if (errMessage == null) {
                Toast.makeText(this@OdexPatchActivity, "Patched to $odexPath", Toast.LENGTH_LONG)
                    .show()
            }
        }
    }
}